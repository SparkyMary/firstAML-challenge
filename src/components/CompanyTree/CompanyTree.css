.companyRoot {
  font-weight: 600;
  margin-bottom: 8px;
}

.tree {
  margin-left: 16px;
}
