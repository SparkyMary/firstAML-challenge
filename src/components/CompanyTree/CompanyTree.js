import React from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import './CompanyTree.css'

import { OWNER_STATUS_TYPES } from '../../constants'

/**
 * Provides the list of all owners of the company with given id
 *
 * @param {String} id - id of the selected company to view the tree
 */
export const getOwners = (id, ownershipArray) => {
  const filteredOwnership = _.filter(ownershipArray, { ownsCompany: id })

  return _.map(filteredOwnership, item => ({
    id: item.id,
    ownsCompany: id,
    nameOfOwner: item.name,
    statusOfOwnship: _.find(OWNER_STATUS_TYPES, { id: item.ownershipType })
      .label,
    ownerId: item.ownerId,
    shares: item.shares
  }))
}

const CompanyTree = ({ nameOfCompany, companyId, ownership }) => {
  const getTreeOfOwners = arrOfItems => {
    return (
      <div>
        {_.map(arrOfItems, item => {
          const subOwners = getOwners(item.ownerId, ownership)
          return (
            <div key={item.id}>
              {item.nameOfOwner} ({item.statusOfOwnship}){' '}
              {item.shares ? `(${item.shares}% Share)` : ''}
              <div className='subOwner'>
                {subOwners.length > 0 && getTreeOfOwners(subOwners)}
              </div>
            </div>
          )
        })}
      </div>
    )
  }

  const owners = companyId ? getOwners(companyId, ownership) : null

  return (
    <div>
      <div className='companyRoot'>{nameOfCompany}</div>
      <div className='tree'>{getTreeOfOwners(owners)}</div>
    </div>
  )
}

CompanyTree.propTypes = {
  nameOfCompany: PropTypes.string,
  companyId: PropTypes.string,
  ownership: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      ownsCompany: PropTypes.string,
      ownershipType: PropTypes.number,
      ownerId: PropTypes.string,
      shares: PropTypes.number
    })
  )
}

export default CompanyTree
