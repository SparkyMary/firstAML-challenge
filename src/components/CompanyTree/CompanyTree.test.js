import React from 'react'
import CompanyTree, { getOwners } from './CompanyTree'
import renderer from 'react-test-renderer'

const ownership = [
  {
    id: '1',
    name: 'B owner',
    ownsCompany: '111',
    ownershipType: 1,
    ownerId: 'BBB',
    shares: 0
  },
  {
    id: '2',
    name: 'C owner',
    ownsCompany: '112',
    ownershipType: 1,
    ownerId: 'CCC',
    shares: 0
  },
  {
    id: '3',
    name: 'D owner',
    ownsCompany: '111',
    ownershipType: 2,
    ownerId: 'DDD',
    shares: 10
  },
  {
    id: '4',
    name: 'E owner',
    ownsCompany: '111',
    ownershipType: 2,
    ownerId: 'EEE',
    shares: 20
  },
  {
    id: '5',
    name: 'F owner',
    ownsCompany: 'BBB',
    ownershipType: 3,
    ownerId: 'FFF',
    shares: 0
  }
]

test('CompanyTree render empty tree', () => {
  const component = renderer.create(
    <CompanyTree
      nameOfCompany={'AAA Company'}
      companyId={'111'}
      ownership={[]}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('CompanyTree renders correctly', () => {
  const component = renderer.create(
    <CompanyTree
      nameOfCompany={'AAA Company'}
      companyId={'111'}
      ownership={ownership}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('HHH', () => {
  const company = { id: '111', name: 'AAA Company' }
  const arrayOfOwners = getOwners(company.id, ownership)
  expect(arrayOfOwners).toHaveLength(3)

  expect(arrayOfOwners[0].nameOfOwner).toEqual('B owner')
  expect(arrayOfOwners[2].shares).toEqual(20)
})
