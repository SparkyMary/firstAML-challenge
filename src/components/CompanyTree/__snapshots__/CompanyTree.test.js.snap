// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyTree render empty tree 1`] = `
<div>
  <div
    className="companyRoot"
  >
    AAA Company
  </div>
  <div
    className="tree"
  >
    <div />
  </div>
</div>
`;

exports[`CompanyTree renders correctly 1`] = `
<div>
  <div
    className="companyRoot"
  >
    AAA Company
  </div>
  <div
    className="tree"
  >
    <div>
      <div>
        B owner
         (
        Director
        )
         
        
        <div
          className="subOwner"
        >
          <div>
            <div>
              F owner
               (
              Trustee
              )
               
              
              <div
                className="subOwner"
              />
            </div>
          </div>
        </div>
      </div>
      <div>
        D owner
         (
        Shareholder
        )
         
        (10% Share)
        <div
          className="subOwner"
        />
      </div>
      <div>
        E owner
         (
        Shareholder
        )
         
        (20% Share)
        <div
          className="subOwner"
        />
      </div>
    </div>
  </div>
</div>
`;
